### PR TITLE
Add semi-support for document apps in Toga GTK

### PR DESCRIPTION
**Pull request by hawkowl**
## PR Checklist:

- [maybe] All new features have been tested
- [uhhh] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: src/gtk/toga_gtk/keys.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+_KEY_MAP = {
+    0xFF51: Key.LEFT,
+    0xFF52: Key.UP,
+    0xFF53: Key.RIGHT,
+    0xFF54: Key.DOWN,
+}
+
+def gtk_to_key(key):
+    return _KEY_MAP.get(key)
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/keys.py]: At column 1: (E302) expected 2 blank lines, found 1

File: src/gtk/toga_gtk/app.py [hunk 2]
```diff
@@ -11,6 +13,7 @@
 from toga.command import GROUP_BREAK, SECTION_BREAK, Command
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py [hunk 3]
```diff
@@ -11,6 +13,7 @@
...
 import toga
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py [hunk 4]
```diff
@@ -11,6 +13,7 @@
...
 from .window import Window
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py [hunk 5]
```diff
@@ -11,6 +13,7 @@
...
+from .dialogs import open_file
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (F401) '.dialogs.open_file' imported but unused

File: src/gtk/toga_gtk/app.py [hunk 6]
```diff
@@ -11,6 +13,7 @@
...
 from toga import Icon
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py [hunk 7]
```diff
@@ -11,6 +13,7 @@
...
 from toga.handlers import wrapped_handler
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/factory.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
 from .app import App, DocumentApp, MainWindow
 from .command import Command
 
+from .document import Document
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/factory.py]: At column 1: (F401) '.document.Document' imported but unused

File: src/gtk/toga_gtk/widgets/webview.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
         pass
 
 from .base import Widget
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/widgets/webview.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/widgets/webview.py [hunk 10]
```diff
@@ -17,6 +17,7 @@
...
+from ..keys import gtk_to_key
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/widgets/webview.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py (hunk 2 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py (hunk 3 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py (hunk 4 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py (hunk 5 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (F401) '.dialogs.open_file' imported but unused

File: src/gtk/toga_gtk/app.py (hunk 6 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/app.py (hunk 7 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/app.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/factory.py (hunk 8 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/factory.py]: At column 1: (F401) '.document.Document' imported but unused

File: src/gtk/toga_gtk/widgets/webview.py (hunk 9 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/widgets/webview.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/widgets/webview.py (hunk 10 above)
Comment (brutusthebee) [inline src/gtk/toga_gtk/widgets/webview.py]: At column 1: (E402) module level import not at top of file

File: src/gtk/toga_gtk/widgets/canvas.py [hunk 11]
```diff
@@ -141,7 +125,7 @@ def write_text(self, text, x, y, font, draw_context, *args, **kwargs):
         elif self.native.font:
             write_font = self.native.font
             write_font.family = self.native.font.get_family()
-            write_font.size = self.native.font.get_size() / SCALE
+            write_font.size = self.native.font.get_size() / Pango.SCALE
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/widgets/canvas.py]: At column 61: (F821) undefined name 'Pango'

File: src/gtk/toga_gtk/widgets/canvas.py [hunk 12]
```diff
@@ -159,7 +143,7 @@ def measure_text(self, text, font, draw_context, *args, **kwargs):
             draw_context.set_font_size(font.size)
         elif self.native.font:
             draw_context.select_font_face(self.native.font.get_family())
-            draw_context.set_font_size(self.native.font.get_size() / SCALE)
+            draw_context.set_font_size(self.native.font.get_size() / Pango.SCALE)
```
Comment (brutusthebee) [inline src/gtk/toga_gtk/widgets/canvas.py]: At column 70: (F821) undefined name 'Pango'

File: src/gtk/toga_gtk/libs/gtk.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
+import gi
+
+gi.require_version('Gtk', '3.0')
+from gi.repository import Gtk, Gio, GLib
+
+# The following import will fail if WebKit or it's API wrappers aren't
```
Comment (bittner) [inline src/gtk/toga_gtk/libs/gtk.py]: "it's" should be "its"

File: src/cocoa/toga_cocoa/keys.py [hunk 14]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    8: Key.C,
+    9: Key.V,
+    11: Key.B,
+    12: Key.Q,
+    13: Key.W,
+    14: Key.E,
+    15: Key.R,
+    16: Key.Y,
+    17: Key.T,
+    18: modified_key(Key._1, shift=Key.EXCLAMATION)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 53: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 15]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    19: modified_key(Key._2, shift=Key.AT)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 44: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 16]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    20: modified_key(Key._3, shift=Key.HASH)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 46: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 17]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    21: modified_key(Key._4, shift=Key.DOLLAR)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 48: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 18]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    22: modified_key(Key._6, shift=Key.CARET)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 47: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 19]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    23: modified_key(Key._5, shift=Key.PERCENT)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 49: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 20]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    24: modified_key(Key.PLUS, shift=Key.EQUAL)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 49: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 21]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    25: modified_key(Key._9, shift=Key.OPEN_PARENTHESIS)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 58: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 22]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    26: modified_key(Key._7, shift=Key.AMPERSAND)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 51: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 23]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    27: modified_key(Key.MINUS, shift=Key.UNDERSCORE)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 55: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 24]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    28: modified_key(Key._8, shift=Key.ASTERISK)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 50: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 25]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    29: modified_key(Key._0, shift=Key.CLOSE_PARENTHESIS)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 59: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 26]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    30: Key.CLOSE_BRACKET,
+    31: Key.O,
+    32: Key.U,
+    33: Key.OPEN_BRACKET,
+    34: Key.I,
+    35: Key.P,
+    36: Key.ENTER,
+    37: Key.L,
+    38: Key.J,
+    39: modified_key(Key.QUOTE, shift=Key.DOUBLE_QUOTE)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 57: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 27]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    40: Key.K,
+    41: modified_key(Key.COLON, shift=Key.SEMICOLON)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 54: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 28]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    42: Key.BACKSLASH,
+    43: modified_key(Key.COMMA, shift=Key.LESS_THAN)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 54: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 29]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    44: modified_key(Key.SLASH, shift=Key.QUESTION)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 53: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 30]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    45: Key.N,
+    46: Key.M,
+    47: modified_key(Key.FULL_STOP, shift=Key.GREATER_THAN)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 61: (F821) undefined name 'event'

File: src/cocoa/toga_cocoa/keys.py [hunk 31]
```diff
@@ -20,101 +21,103 @@ def mod_fn(modifierFlags):
...
+    48: Key.TAB,
+    49: Key.SPACE,
+    50: modified_key(Key.BACK_QUOTE, shift=Key.TILDE)(event.modifierFlags),
```
Comment (brutusthebee) [inline src/cocoa/toga_cocoa/keys.py]: At column 55: (F821) undefined name 'event'

**Pull request merged**